libs/vkd3d/command.c

```diff
@@ -11891,6 +12218,93 @@ static void d3d12_command_queue_signal(struct d3d12_command_queue *command_queue
     /* We should probably trigger DEVICE_REMOVED if we hit any errors in the submission thread. */
 }
 
+static void d3d12_command_queue_wait_shared(struct d3d12_command_queue *command_queue,
+        struct d3d12_shared_fence *fence, UINT64 value)
+{
+    assert(fence->timeline_semaphore);
+    vkd3d_queue_add_wait(command_queue->vkd3d_queue, &fence->ID3D12Fence_iface, fence->timeline_semaphore, value);
+
+    /* We can't get the physical value here for a shared fence,
+     * and waiting for the next submit would add even more time
+     * for a fence rewind to go unnoticed, so we flush. */
+    vkd3d_queue_flush_waiters(command_queue->vkd3d_queue, &command_queue->fence_worker, &command_queue->device->vk_procs);
```

Comment:
Is there idea here that shared fences work like D3D12 timelines and they can rewind and everything and winevulkan resolves this on submit internally instead?

Does vkQueueSubmitKHR here basically translate into a blocking wait if it finds a timeline semaphore that is D3D12 shared?

Reply:
Yeah, that's the idea, but vkQueueSubmitKHR won't block, there's another queue worker thread inside winevulkan, similar to VKD3D-Proton's.  The only difference really is that winevulkan uses a mechanism to start the wait inside the vkQueueSubmitKHR call, in order to avoid missing any values that would signal that would then be rolled back by a signal.

If you'd like to see what that's like, here's my wine branch based off Proton Experimental: https://github.com/Guy1524/wine/commits/halo-infinite-fixes

Reply:
Ok, that seems fine.